src/test/java/server/EchoServerTest.java

```diff
@@ -1,12 +1,65 @@
 package server;
 
+
+import mocks.MockWrapper;
 import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+
 import static org.junit.jupiter.api.Assertions.*;
 
 
-class EchoServerTest {
-    @Test void walkingSkeletonTest() {
-        EchoServer classUnderTest = new EchoServer();
-        assertEquals(classUnderTest.walkingSkeletonTest(), "testing 123");
+class MockWrappersTests {
+    @Test
+    public void testRunGetsAcceptsClient() throws IOException {
+        ServerSocket serverSocket = new ServerSocket(4000);
+        MockWrapper mockWrapper = new MockWrapper(serverSocket);
+        EchoServer testServer = new EchoServer();
+
+        testServer.run(mockWrapper);
+
+        assertEquals(1, mockWrapper.getNumberOfCallsToAcceptClient());
+    }
+    @Test
+    public void testRunCreatesInputStream() throws IOException {
+        ServerSocket serverSocket = new ServerSocket(4444);
+        MockWrapper mockWrapper = new MockWrapper(serverSocket);
+        EchoServer testServer = new EchoServer();
+
+        testServer.run(mockWrapper);
+
+        assertEquals(1, mockWrapper.getNumberOfCallsToCreateInputStream());
+    }
+    @Test
+    public void testRunCreatesOutputStream() throws IOException {
```

Comment:
**Danger: Pointy-headed Programming Language Nerd Musings Ahead**

(Just think about this, no need to do anything in particular)

Another thought that I had for all the tests as a whole, is how much can you lean on the type system? One of the advantages of Java, and static typing in general, is that the type system is _effectively writing unit tests for you_. The sort of things where in ruby you need to check that something indeed returns a certain type, or even a certain implicit interface (duck type), aren't really needed in Java and other statically-typed languages. In fact, if you try to write a negative test for this sort of thing, you can't:

```java
class Main {  
  public static void main(String args[]) {
    int result = foo();
    if (result == "Oops") {
      System.err.println("OK");
    } else {
      System.err.println("FAIL");
    }
  }

  static String foo() { return "Oops"; }
}
```

I'm trying to write a (clumsy) test for `foo()` but I can't even write it:

```
Main.java:5: error: incompatible types: String cannot be converted to int
    int result = foo();
                    ^
Main.java:6: error: bad operand types for binary operator '=='
    if (result == "Oops") {
               ^
  first type:  int
  second type: String
2 errors
```

The compiler says that the code is [not even wrong](https://en.wikipedia.org/wiki/Not_even_wrong). If you do it right, you can sort of jiu-jitsu the complier into making code that's logically wrong into code that won't even compile. That way there's nothing to maintain and the bad code also can't gradually creep back into the codebase. There's a idea that goes by ["making illegal states unrepresentable" (deep dive)](https://ybogomolov.me/making-illegal-states-unrepresentable/) that tries to do this maneuver. Or if you are bored and want a video to watch, this is [a good one](https://www.youtube.com/watch?v=IcgmSRJHu_8).